### ISSUE TITLE
Don't save to the program folder on Windows

### ISSUE THREAD

**Issue by colejohnson66** [project64/project64#1353]
Programs should not be writing to their own "Program Files" folder. They should be saving to `%LOCALAPPDATA%`, or somewhere else in the User's directory. Maybe `%USERPROFILE%/Project64`?

**Comment by Frank-74** [comment]
One solution I can think of is an option in the first run wizard, to choose the directory for Config and Save folders.

There's no problem with writing to the Save folder under Program Files.  As long as the install has been run.  The installer sets the folder permissions for write access.

I made changes to the installer a long while ago, to set full permission to the Project64.exe folder and directories below it.  It allowed 1.6 plugins to work correctly within Program Files on Windows 8/10.  1.6 plugins write to an RDB in the Project64.exe folder.  If permission isn't set for the root folder, they don't save their settings.

**Comment by project64** [comment]
the proper place might be local app data or user profile, but the problem is that a lot of user find it hard to find. Also if you are using it in its own dir, those files will be shared between different instances

**Comment by Frank-74** [comment]
@project64 
> problem is that a lot of user find it hard to find.

Revamp the directories options, it needs updating, add a button to open folder for each directory.
Give the option to choose/edit the default Config/Save directory on first run wizatd, apply the changes to the directories options

Just thinking out loud...

**Comment by shawnz** [comment]
Please fix this issue, it is not appropriate for security reasons to give unprivileged users write access to directories underneath the program files directory. This lets unprivileged users, for example, bypass storage limits on their profile, store malware where administrators might accidentally run it, etc. Consider the following scenario: An administrator installs Project64 on their computer. An unprivileged user edits the config file (which is writable by anyone) to point the plugins directory somewhere they control. Then they put some malware in place of the real plugins. Now when the administrator opens Project64, the attacker's malware will run with administrative privileges.

As far as the issue of AppData being too hard to find, you could store the data in My Documents instead if you want the user to be able to see and manage the files with windows explorer. AppData is meant for storing "hidden" data, not files which the user is supposed to be able to see and manage themselves.

**Comment by JunielKatarn** [comment]
Why not make a command line arg for `Project64.exe` to indicate what's the "local" directory?
i.e. `Project64.exe -UserDir %USERPROFILE%\Desktop\pj64data`

**Comment by project64** [comment]
I need to think more how to do this better .. the structure is correct and good and I do not want to change it if it not in program files directory.

now if it is in program files directory there is an argument that it should be else where, now those directories are hidden, and there could be methods to find it.

the issue that @shawnz raise is possible but highly unlikely. You would have to have malicious dll's, access to the machine. The admin user would have to manually run it as admin as it executes as non privileged user on everything since XP.  

Maybe the best solution is just remove installer and have a zip instead. 

lots of ok solutions, nothing I can choose that is a clear obvious solution

**Comment by shawnz** [comment]
Thanks zilmar for all your hard work.

For now as a workaround I noticed the installer has a "portable mode" option, so that is basically as good as offering a zip file. However even in portable mode it still sets the permissions for Everyone so that is something to keep in mind for anyone using this technique.

**Comment by Jj0YzL5nvJ** [comment]
> Maybe the best solution is just remove installer and have a zip instead.

 That is the most reasonable solution, most plugins save their own settings, shaders, etc. inside the PJ64 folder or the registry anyway.
 Installing PJ64 has never been functional in the strict NT architecture sense, and unless you put end to the 'plugin hell' or PJ64 start copying the RetroArch style, that will never change.
 Plus, by dropping the installer you can rule out issues generated indirectly by the [UAC virtualization](https://en.wikipedia.org/wiki/User_Account_Control#Features).

**Comment by project64** [comment]
the installer was meant to make it easier on end users .. tho it was designed in an XP era.

maybe with 2.4 I can give the option of installer and zip file and see what the response is. If the majority use zip, then I can drop the installer. If the majority use the installer then I am back to this original problem.

**Comment by Jj0YzL5nvJ** [comment]
Even in NT/2K/XP, using an administrative PJ64 installation, in a non-administrative account there is the same problem. UAC is a bad patch attempt for "Layer 8", many users used XP in a Win9X sense instead of a NT sense.

If you really want "fix" PJ64 for "admins" (not worth it) look to mpv:
https://mpv.io/manual/master/#configuration-files
https://mpv.io/manual/master/#files
https://mpv.io/manual/master/#files-on-windows

**Comment by oddMLan** [comment]
I second the motion of getting rid of the installer. I never found the installer particularly convenient, as I keep emulators in a dedicated folder personally. And non-admin users will have trouble installing the emulator as it'll ask for admin credentials, unless they use the Portable option in the installer, which might not be obvious to all. Unzipping is way easier and quicker, in my opinion.

**Comment by dftf-stu** [comment]
Yes, testing both the v2.3-210 and v3.0.1-5664 installers (the latest versions available [in the Public Releases](https://www.pj64-emu.com/public-releases)) section, I find that they default to the following folders:
- **Plug-Ins:** C:\Program Files\Project64 3.0\Plugin\
- **Saves:** C:\Program Files\Project64 3.0\Save\
- **Saved states:** C:\Program Files\Project64 3.0\Save\
- **Screenshots:** C:\Program Files\Project64 3.0\Screenshots\
- **Texture Packs:** C:\Program Files\Project64 3.0\Saves\

On **Windows XP**, this would mean any non-administrator would be unable to have their settings for the app remembered, or saves to work, as by-default they cannot write to that folder.

On **Windows Vista** or later, where UAC has not been disabled, then "File and Registry Virtualisation" would kick-in and relocate the folders.  For example
• C:\Program Files (x86)\Project64 3.0\Screenshots
would get relocated to:
• C:\Users\Joe.Bloggs\AppData\Local\VirtualStore\Program Files (x86)\Project64 3.0\Screenshots

It would be better to either save the relevant data types under the "AppData" folders:
- **Plug-Ins:** C:\Users\Joe.Bloggs\AppData\Local\Project64\4.0\PlugIns
- **Saves:** C:\Users\Joe.Bloggs\AppData\Roaming\Project64\4.0\Save
- **Saved states:** C:\Users\Joe.Bloggs\AppData\Roaming\Project64\4.0\Save
- **Screenshots:** C:\Users\Joe.Bloggs\AppData\Local\Project64\4.0\Screenshots
- **Texture Packs:** C:\Users\Joe.Bloggs\AppData\Local\Project64\4.0\Textures

Or just simply do what other emulators, such as DuckStation do, and just create a hidden "Project64" inside the user's "Documents" folder (C:\Users\Joe.Bloggs\Documents\Project64) and save everything inside there.

**Comment by project64** [comment]
some users find dowloading an installer, clicking install and then run a lot more easier.  This I am guessing is the vast majority of the casual users which is why I have the installer. 

I know when vista started to redirect to AppData dir it caused lots of issues cause those directories are generally hidden from most users.

Because of this and that then if you do have multiple users it will use that directory, I will never use AppData directory, I want everything to stay together.

**Comment by DerekTurtleRoe** [comment]
I can look into making the installer more modern and user-friendly at some point if you'd like, although I would say getting Project64 working well in a portable fashion first and then making the installer just "the portable version but installed" might work best as an approach.

**Comment by dftf-stu** [comment]
@DerekTurtleRoe: given _Project64_ does not have any sort of auto-updater, nor seems to require any "dependencies" (to use the _Linux_ term), such as a _Visual C++_ library, then I struggle to see the value in having an installer anyway.

The whole-point of an installer version of an app usually is that it will install any extras an app needs (e.g. Visual C++ 2015-2019 or .NET Framework 4.8), and usually sets-up an auto-updater service or Scheduled Task.  If _Project64_ does neither, I don't really see the value, given the ability to extract a ZIP file has been present in Windows natively now since _2000_ and _Me_...

**Comment by flagrama** [comment]
Many people literally don't know how to unarchive a zip, but know how to run an installer. They choose the one they understand.

**Comment by Jj0YzL5nvJ** [comment]
> Many people literally don't know how to unarchive a zip, but know how to run an installer.

Yes, and exactly the same people run a setup of a double layer DVD of an ISO from WinRAR and then complain that the installer or the installed program does not work or does not do what it should do.

Things get interesting when the installer asks for the disk 2.
